src/Domain/ElementLogicalCoordinates.cpp

```diff
@@ -0,0 +1,109 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "ElementLogicalCoordinates.hpp"
+
+#include <array>
+#include <memory>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "DataStructures/Tensor/Tensor.hpp"  // IWYU pragma: keep
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Domain/BlockLogicalCoordinates.hpp"  // IWYU pragma: keep
+#include "Domain/ElementId.hpp"                // IWYU pragma: keep
+#include "Domain/Side.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/MakeArray.hpp"
+
+template <size_t Dim>
+std::unordered_map<ElementId<Dim>, ElementLogicalCoordHolder<Dim>>
+element_logical_coordinates(const std::vector<ElementId<Dim>>& element_ids,
+                            const std::vector<BlockLogicalCoordHolder<Dim>>&
+                                block_coord_holders) noexcept {
+  // Temporarily put results here in data structures that allow
+  // push_back, because we don't know the sizes of the output
+  // DataVectors ahead of time.
+  std::vector<std::array<std::vector<double>, Dim>> x_element_logical(
+      element_ids.size());
+  std::vector<std::vector<size_t>> offsets(element_ids.size());
+
+  // Loop over points
+  for (size_t offset = 0; offset < block_coord_holders.size(); ++offset) {
+    const size_t block_id = block_coord_holders[offset].block_id;
+    const auto& x_block_logical =
+        block_coord_holders[offset].block_logical_coords;
+    // Need to loop over elements, because the block doesn't know
+    // things like the refinement_level of each element.
+    for (size_t element_index = 0; element_index < element_ids.size();
+         ++element_index) {
+      const auto& element_id = element_ids[element_index];
+      if (element_id.block_id() == block_id) {
+        // This element is in this block; now check if the point is in
+        // this block.
+        bool is_contained = true;
+        auto x_elem = make_array<Dim>(0.0);
+        // std::array<double, Dim> x_elem;
+        for (size_t d = 0; d < Dim; ++d) {
+          const double up =
+              gsl::at(element_id.segment_ids(), d).endpoint(Side::Upper);
```

Comment:
I don't know, but is it possible for segments_ids to be flipped such that e.g. +1 is the Lower endpoint and -1 is the Upper endpoint?

Reply:
@gsb76 no this isn't possible as all elements within a block are aligned